src/ringct/rctSigs.cpp

```diff
@@ -1578,6 +1581,61 @@ namespace rct {
       }
     }
 
+    static struct SRctCache
+    {
+      mutable std::mutex m;
+      std::unordered_set<crypto::hash> hashes;
+      crypto::hash buf[8192] = {};
+      uint32_t counter = 0;
+
+      template<typename T>
+      static crypto::hash get_hash(const T& data)
+      {
+        std::stringstream ss;
+        binary_archive<true> ar(ss);
+        ::do_serialize(ar, const_cast<T&>(data));
+
+        crypto::hash h;
+        cryptonote::get_blob_hash(ss.str(), h);
+
+        return h;
+      }
+
+      void add(const crypto::hash& h)
+      {
+        std::lock_guard<std::mutex> lock(m);
+        if (hashes.insert(h).second)
+        {
+          crypto::hash& old_hash = buf[counter++ % 8192];
+          hashes.erase(old_hash);
+          old_hash = h;
+        }
+      }
+
+      bool find(const crypto::hash& h) const
+      {
+        std::lock_guard<std::mutex> lock(m);
+        return (hashes.find(h) != hashes.end());
+      }
+    } rct_cache;
+
+    bool verRctNonSemanticsSimple(const rctSig & rv)
+    {
+      // Don't cache older (or newer) rctSig types
+      if (rv.type != RCTTypeCLSAG && rv.type != RCTTypeBulletproofPlus)
```

Comment:
Instead of making it a positive check (i.e. must have type X or Y), maybe make it a negative check (i.e. type != one of the old sig types)

Reply:
I think it's better to forget to activate this check than to have it running silently with new RCT types and people forgetting that it exists. I can add an assert if `rv.type` is some new unknown type.

Reply:
Also, this assert will only be needed for this PR in master branch, not for the release branch because it won't add any new types.

Reply:
> I think it's better to forget to activate this check than to have the cache running silently with new RCT types and people forgetting that it exists. I can add an assert if `rv.type` is some new unknown type.

Why's that? Are you worried about the security of the cache in future updates? 

Also, the assertion statement in the latest commit you pushed might have made the situation worse... now core consensus code will error out if we forget to update this part of the code.

Reply:
I don't know how future updates will work and no one knows right now. Maybe they will change how data is serialized and this code will become incorrect. No, this assert makes sure this code won't be forgotten because it will block all transactions with new RingCT types, so it will be taken care of during development.

Reply:
I'd be down to add this assert statement is as long as it gets added into the update checklist.

Reply:
Either it will trigger during development, and it can be fixed then, or it won't, because there's no problem. That's kind of the point of adding an assert.

Reply:
What update checklist? It's impossible that the release process would get that far without this code being looked at. You still don't get how this assert will work.

Reply:
> Either it will trigger during development, and it can be fixed then, or it won't, because there's no problem. That's kind of the point of adding an assert.

Yeah it will get triggered eventually, but I would prefer if it were to be thought about sooner rather than later if a new signature type were to be added. It would be nice to have something additional besides just a runtime assertion. Even if it's only just a comment in `rctTypes.h` where `RCTType*` are defined, so the number of places in the code code where updates are needed doesn't grow undocumented.

Reply:
There are dozens places in the codes where RCTType* is checked and used, and zero comments about it in `rctTypes.h`. The assert I added is a standard practice used in Monero code already, example: https://github.com/monero-project/monero/blob/master/src/cryptonote_basic/cryptonote_format_utils.cpp#L464
Whoever will be working on adding a new RCT type, *will* search the code for all occurences of RCTType* and fix all places where it's found - even before this assert triggers. This assert is just an extra safety check that will make it impossible to send/receive transactions until it's fixed, so it's literally impossible it would pass any tests without a fix. Why do I have to explain this, you should know this all. This issue is resolved already.

Reply:
> There are dozens places in the code where RCTType* is checked and used, and zero comments about it in `rctTypes.h`

I think this more of an example of "it is what it is" instead of "what it should be", but just to be clear, I'm still in favor of the PR besides this, and am willing to concede this point if no one else cares.